File: Swift/LeetCode/LeetCode/LongestCommonPrefix.swift
//
//  LongestCommonPrefix.swift
//  LeetCode
//
//  Created by 黎赵太郎 on 25/11/2017.
//  Copyright © 2017 lizhaotailang. All rights reserved.
//
//  Write a function to find the longest common prefix string amongst an array of strings.
//
//  Accepted. See [LongestCommonPrefixTests](./LeetCodeTests/LongestCommonPrefixTests.swift) for test cases.
//

import Foundation

class LongestCommonPrefix {
    
    func longestCommonPrefix(_ strs: [String]) -> String {
        if strs.isEmpty {
            return ""
        }
        
        if strs.count == 1 {
            return strs[0]
        }
        
        var allChars: [[Character]] = []
        for s in strs {
            if s.isEmpty {
                return ""
            }
            allChars.append(Array(s))
        }
        
        for i in 0...(allChars[0].count - 1) {
            for j in 1...(allChars.count - 1) {
                if i == allChars[j].count || allChars[0][i] != allChars[j][i] {
                    return String(allChars[0][0..<i])
                }
            }
        }
        
        return String(allChars[0])
    }
    
}

File: Golang/LongestCommonPrefix/LongestCommonPrefix.go
package LongestCommonPrefix

//Write a function to find the longest common prefix string amongst an array of strings.
//
//Accepted.

func longestCommonPrefix(strs []string) string {
	if strs == nil || len(strs) == 0 {
		return ""
	}

	if len(strs) == 1 {
		return strs[0]
	}

	for i := 0; i < len(strs[0]); i++ {
		for j := 1; j < len(strs); j++ {
			if i == len(strs[j]) || strs[0][i] != strs[j][i] {
				return string(strs[0][0:i])
			}
		}
	}
	return strs[0]
}
